### PR TITLE
fixes #18329 - adds LVM snapshots to backup

### PR DESCRIPTION
**Pull request by cfouant**
To test:
Set up a katello instance on a LVM filesystem (cannot use the Vagrant set up). You may create partitions separately, such as mounting /var/lib/pulp separate from root. The volume group must have enough space to hold a snapshot, which by default would be about 6G.

### CODE REVIEW

File: katello/katello-backup
Comment (sean797) [inline katello/katello-backup]: Small thing but I think running `--exclude foo,bar` is kind of unnecessary here

File: katello/katello-backup
Comment (sean797) [inline katello/katello-backup]: What if my LV is named something other than root?
What if I have separate partitions for other mount points?

File: katello/katello-backup
Comment (sean797) [inline katello/katello-backup]: `ERROR: Snapshot backups cannot be done online or incrementally` ?

Also, why can't we do them online?

**Review by sean797 — CHANGES_REQUESTED** [review]
TBH I don't think this fits in our scripts. 

If you want to use LVM snapshots its only 1 cmd per mount point, only the user knows the mount points and is going to be wildly different on each install.

IMO its kind of pointless to essentially use katello-backup to wrap `lvcreate -s..... `

Also what about katello-restore?

File: katello/katello-backup
Comment (cfouant) [reply katello/katello-backup]: I'm pretty new to logical volumes, so I'm sure my knowledge base is missing some information. I was under the impression that to take a snapshot of root, which is where all the files we look to backup are located, it would be in `/dev/<volume group name>/root`. Is this incorrect?

File: katello/katello-backup
Comment (cfouant) [reply katello/katello-backup]: Respectfully disagree on this point. Since we aren't stopping the excluded services, we shouldn't attempt to start them, if only for the few seconds of time it may save. If the excluded services should grow, this keeps it more manageable and obvious as well.

File: katello/katello-backup
Comment (cfouant) [reply katello/katello-backup]: I was under the impression that the snapshot must be created while services were shut off. If that isn't the case, I'm happy to add an online solution. I'll work on updating this one now.

File: katello/katello-backup
Comment (evgeni) [inline katello/katello-backup]: `-txfs` implies the user has XFS, which they don't have to. I would just omit passing `-t` and let `mount` handle it automatically.

**Comment by cfouant** [comment]
@thomasmckay @bbuckingham - are there any members of your team available to review this? I can help on the setup if needed, but it would be great if there were any who already understood lvm filesystem so if I have errors in my knowledge base, I'm not passing on misinformation!

File: katello/katello-backup
Comment (evgeni) [inline katello/katello-backup]: * this will fail if there is already a snapshot called `rootsnapshot` (which the user could have created them-self for some reason)
* this implies that all data is on the same LV with the name `root`. a prod deployment usually will have `/var`, `/var/lib/pgsql/`, `/var/lib/mongodb` and `/var/lib/pulp` on separate LVs, with varying names.

Comment (evgeni) [reply katello/katello-backup]: Also we should probably check if the VG actually has enough free space, even before stopping the services

**Review by evgeni — CHANGES_REQUESTED** [review]

**Comment by evgeni** [comment]
I think the snapshotting logic should be more like this (pseudo code, but I think you get the idea):

```ruby
snaps = []
DATABASES.each do |db|
  lv = find_lv_for_mountpoint(db.mountpoint)
  snap = lv.create_snapshot
  snaps << snap
end
backup
snaps.each do |snap|
  snap.umount
  snap.remove
end
```

File: katello/katello-backup
Comment (evgeni) [inline katello/katello-backup]: VG location seems to imply a path, so I tried `--snapshot /dev/VolGroup00` and it failed. I think it should say "VG name" here? And we could check for the name not having any `/` in it to be sure the user gave a name, not a path.

**Comment by evgeni** [comment]
That failed in my test VM with

```
Backing up mongo db... 
tar: var/lib/mongodb/journal/prealloc.0: File shrank by 1071837184 bytes; padding with zeros
tar: var/lib/mongodb/journal/prealloc.1: Read error at byte 0, while reading 2048 bytes: Input/output error
tar: var/lib/mongodb/journal/prealloc.2: Read error at byte 0, while reading 4608 bytes: Input/output error
tar: Exiting with failure status due to previous errors
Failed 'tar --selinux --create --file=/var/tmp/katello-backup/katello-backup-20170509174658/mongo_data.tar --listed-incremental=/var/tmp/katello-backup/katello-backup-20170509174658/.mongo.snar --exclude=mongod.lock var/lib/mongodb/'
```
I think this happened because the snapshot I created was too small, so we might want to enforce a certain minimum size (esp as I was writing the backup to the same LV as the snapshot was, which seems like a terrible idea).

**Comment by evgeni** [comment]
Another thing: I would have thought that `--snapshot` and `--online-backup` are mutually exclusive as you want to have a snap of the on-disk data, which does not seem possible while the services are online?

**Review by evgeni — CHANGES_REQUESTED** [review]

File: katello/katello-backup
Comment (evgeni) [inline katello/katello-backup]: For some reason it created `/rootsnapshot/` for me, and not `/mnt/snap/rootsnapshot/`.

**Comment by evgeni** [comment]
Some brain dump from an offline discussion with @cfouant in the office:
* snapshot and online should be exclusive, as otherwise we would end up with a live online backup of pgsql and mongo (as we use export commands, talking to a running daemon, which reads the non-snapshotted data) and an old backup of pulp (as it uses the data from the snapshot). this pretty sure makes the whole backup inconsistent and error prone
* we think that katello-restore should not need any adjustments, as the prereq for a restore is a default install of katello, thus all mountpoints/LVs/etc should be already preset and are up to the user (and thus its not any change to the old behavior)
* as we can't possibly know how the user has their LVs set up, we should probably just ask the user for a list of mountpoints that are on separate LVs and use that list for the snapshots.

@cfouant - did I miss anything?
@sean797 - does the above make sense to you?

**Comment by sean797** [comment]
@cfouant @evgeni most of the above makes sense though it does lead me to the same conclusion as before.

This is how I see it:
- I don't think we should try to become an all singing, all dancing backup script with loads of features/options. There are plenty of other tools dedicated to this that will do a better job.
   - If we add LVM snapshots, should we add rsync, Relax-and-Recover, backuping up to ISOs, where does this stop ?
- If we need to ask the user for every option we pass to `lvcreate -s` then what benefit does this actually have? What's the point in a wrapper script if it doesn't do anything cleaver?
- There are many possible options valid for `lvcreate -s`. What if I want to use ` --thinpool`, `--extents` or others, do we accept values for all these options? Where does this end?
- An inexperienced sysadmin with knowledge of their LVs set up could script backups using LVM snapshots in 15 minutes and learn something along the way.
- Adds complexity to this script

Sorry if I sound super negative about this, I just see katello-backup as a simple, ~~quick~~ & cheap backup.

I see this as a DOC bug, people need to know what directories they need to backup if they would like to use other tools, `katello-backup list` to list directories requiring backup would suffice IMO.

**Comment by cfouant** [comment]
@sean797 - I definitely see where you're coming from, and agree a lot in limiting features; however, snapshots are a wildly anticipated feature that's definitely going to happen. I was able to limit snapshots to just LVM, which I mention because there were requests for 3 different snapshot options!

File: katello/katello-backup
Comment (evgeni) [inline katello/katello-backup]: this should be `@databases = DATABASES.dup`, otherwise:

```ruby
irb(main):001:0> FOO = [1,2,3]
=> [1, 2, 3]
irb(main):002:0> @foo = FOO
=> [1, 2, 3]
irb(main):003:0> @foo.delete 1
=> 1
irb(main):004:0> @foo
=> [2, 3]
irb(main):005:0> FOO
=> [2, 3]
```

File: katello/katello-backup
Comment (evgeni) [inline katello/katello-backup]: I don't this we need the fstype, just ommit it and let the kernel handle it for you

File: katello/katello-backup
Comment (evgeni) [inline katello/katello-backup]: I don't this we need the fstype, just ommit it and let the kernel handle it for you

Comment (evgeni) [inline katello/katello-backup]: please don't hardcode the VG as `rhel`, users will have various names here. `#{snapshot_location}` should be exactly what you need.

File: katello/katello-backup
Comment (evgeni) [inline katello/katello-backup]: `findmnt -n -o SOURCE --target #{target}`

`-o SOURCE` outputs only the source column, so you don't have to `split()`

File: katello/katello-backup
Comment (evgeni) [inline katello/katello-backup]: `File.join`? :)

**Review by evgeni — CHANGES_REQUESTED** [review]

File: katello/katello-backup
Comment (cfouant) [reply katello/katello-backup]: However, if the file type is xfs, there needs to be the -onouuid,ro tag added. Without it, mounting fails.

File: katello/katello-backup
Comment (cfouant) [reply katello/katello-backup]: That was an accident/old code, thanks for seeing it!

File: katello/katello-backup
Comment (cfouant) [reply katello/katello-backup]: I do love that new trick you taught me :) will change

File: katello/katello-backup
Comment (cfouant) [reply katello/katello-backup]: AWESOME, thanks

File: katello/katello-backup
Comment (cfouant) [reply katello/katello-backup]: Noted, will change

File: katello/katello-backup
Comment (evgeni) [reply katello/katello-backup]: Mhh. Right.
I'd even argue that `-o ro` should *always* be added.

File: katello/katello-backup
Comment (evgeni) [inline katello/katello-backup]: Heh, I fear now I talked you into something that won't work either :(
`nouuid` is specific for xfs, so we can't pass it unconditionally. `ro` is OK for all fstypes, though.

File: katello/katello-backup
Comment (evgeni) [inline katello/katello-backup]: that creates snapshots the size of the current disk usage. that means if I have 600G pulp, I need 600G free space for the snapshot. ain't nobody has so much space left :(

File: katello/katello-backup
Comment (evgeni) [inline katello/katello-backup]: this creates one snapshot per database, in the case where multiple dbs are on the same LV, we thus create multiple snaps of the same data. not too bad (we do that in the "split everything" case too) but one could safe a bit of space here. would make the code more complicated, though

**Review by evgeni — CHANGES_REQUESTED** [review]

**Comment by cfouant** [comment]
[test]

File: katello/katello-backup
Comment (evgeni) [inline katello/katello-backup:109]: above you say the default is 5G, what is correct? ;)

File: katello/katello-backup
Comment (evgeni) [inline katello/katello-backup]: This errors out with:
```
Creating and mounting pulp snapshot
katello-backup:125:in `block in create_and_mount_snapshots': undefined method `get_snapshot_location' for main:Object (NoMethodError)
	from katello-backup:118:in `each'
	from katello-backup:118:in `create_and_mount_snapshots'
	from katello-backup:112:in `snapshot_backup'
	from katello-backup:271:in `<main>'
```

**Review by evgeni — CHANGES_REQUESTED** [review]

File: katello/katello-backup
Comment (cfouant) [reply katello/katello-backup]: @evgeni - ah, i fixed it, that method didn't make it in to the script i pushed. Please try again, working now!

**Comment by evgeni** [comment]
Current version fails for me with:

```
Backing up mongo db... 
tar: journal/j._0: file changed as we read it
Failed 'tar --selinux --create --file=/dev/shm/backup/katello-backup-20170613064103/mongo_data.tar --listed-incremental=/dev/shm/backup/katello-backup-20170613064103/.mongo.snar --exclude=mongod.lock --transform 's,^,var/lib/mongodb/,S' -S *'
```

This is due to https://github.com/Katello/katello-packaging/pull/414#discussion_r121596736

File: katello/katello-backup
Comment (evgeni) [inline katello/katello-backup]: the database name in the array is `mongodb`, not `mongod`

File: katello/katello-backup
Comment (evgeni) [inline katello/katello-backup]: the database name in the array is `mongodb`, not `mongod`

File: katello/katello-backup
Comment (evgeni) [inline katello/katello-backup]: this loop will make the function return `nil` when when the target is not the last item in the list. I would suggest using something like this:

```ruby
  result = nil
  Find.find(File.join(@mountdir, database)) do |path|
    result = File.dirname(path) if File.basename(path) == target
  end
  result
```

**Review by evgeni — CHANGES_REQUESTED** [review]

File: katello/katello-backup
Comment (cfouant) [reply katello/katello-backup:109]: Updated

File: katello/katello-backup
Comment (cfouant) [reply katello/katello-backup]: ah, great catch, updating now!

**Comment by evgeni** [comment]
ACK on this PR from me. @ehelms @johnpmitsch @sean797 @chris1984 what do you all think?

However, I think we should prioritize the following followups:
* add tests for backup/restore (via bats in forklift?)
* add error handling to at least restore katello services after the backup failed, as currently a failed backup will also leave the services down :/

**Review by evgeni — APPROVED** [review]

**Comment by cfouant** [comment]
@evgeni - there's a separate PR for error handling, if you want to review that! https://github.com/Katello/katello-packaging/pull/451

@johnpmitsch has started on a test suite, the project is in the beginning stages, but he is creating a PR when it's come a little further

**Comment by johnpmitsch** [comment]
@evgeni I started working on some ansible playbooks that will perform different backup/restore scenarios on a katello instance, I have there [here](https://github.com/johnpmitsch/katello-backup-restore-test) for now, once they are reliable we can talk with @ehelms about working them into a CI/CD system. I would rather test on the PR than in bats, but we can discuss this more. For now, they are something to run locally to test PRs.

**Review by johnpmitsch — APPROVED** [review]
I didn't test the functionality (I'm assuming @evgeni has), but the code looks fine to me

**Comment by cfouant** [comment]
[test]

**Comment by sean797** [comment]
By the way I'm not very happy with this being merged 20 minutes after the first ACK, especially considering the comments I made in https://github.com/Katello/katello-packaging/pull/414#issuecomment-300616907, I left my 'requesting changes' flag set and the only reply being:
> agree a lot in limiting features; however, snapshots are a wildly anticipated feature that's definitely going to happen.

Saying something is "definitely going to happen." is not valid. I also don't seen any evidence of any users requesting this, my search on foreman-users yields nothing.

I was actually planning on taking another look at this when @cfouant was happy with it to check all my previous points were still valid and actually get a response other than its "definitely going to happen".

But whatever. Just in future we should try to counter people views with facts and/or opinions & allow more than 20 minutes before being merged, especially when someone has expressed another opinion or view.

**Comment by cfouant** [comment]
@sean797 - I did request feedback from users on a satellite6 email thread, and got few responses. I am not sure if you're on that list or not. There are 4 open customer cases regarding using snapshots in BZ. I also responded to your comment on the 15th, and this merged a month after that, with no feedback from you in that interim.

**Pull request merged**
